openssl/src/ssl/bio.rs

```diff
@@ -17,19 +18,38 @@ pub struct StreamState<S> {
     pub panic: Option<Box<Any + Send>>,
 }
 
-pub fn new<S: Read + Write>(stream: S) -> Result<(*mut BIO, Arc<BIO_METHOD>), SslError> {
-    let method = Arc::new(BIO_METHOD {
-        type_: BIO_TYPE_NONE,
-        name: b"rust\0".as_ptr() as *const _,
-        bwrite: Some(bwrite::<S>),
-        bread: Some(bread::<S>),
-        bputs: Some(bputs::<S>),
-        bgets: None,
-        ctrl: Some(ctrl::<S>),
-        create: Some(create),
-        destroy: Some(destroy::<S>),
-        callback_ctrl: None,
-    });
+/// Safe wrapper for BIO_METHOD
+pub struct BioMethod<S> {
+    method: BIO_METHOD,
+    _p: PhantomData<S>
+}
+
+impl<S> BioMethod<S>
+    where S: Read + Write
+{
+    pub fn new() -> BioMethod<S> {
+        BioMethod {
+            method: BIO_METHOD {
+                type_: BIO_TYPE_NONE,
+                name: b"rust\0".as_ptr() as *const _,
+                bwrite: Some(bwrite::<S>),
+                bread: Some(bread::<S>),
+                bputs: Some(bputs::<S>),
+                bgets: None,
+                ctrl: Some(ctrl::<S>),
+                create: Some(create),
+                destroy: Some(destroy::<S>),
+                callback_ctrl: None,
+            },
+            _p: PhantomData
+        }
+    }
+}
+
+unsafe impl<S: Send> Send for BioMethod<S> {}
```

Comment:
In both cases we don't have an unsafe impl on SslStream. In they layout above, we make BioMethod non-generic and unconditionally Send since all it is is a struct of function pointers. The only thing in an SslStream that may cause us to be !Send is S, which is taken care of by the _p field. Does that make sense?